gateway/context.go

```diff
@@ -0,0 +1,257 @@
+	if rc.presetFailureHTTPResponse != nil {
+		rc.writeHTTPResponse(rc.presetFailureHTTPResponse, w)
+		return
+	}
+
+	// Processing a request only gets to this point if a QoS instance was matched to the request.
+	// Use the QoS context to obtain an HTTP response.
+	// There are 3 possible scenarios:
+	// 	1. The QoS instance rejected the request, e.g. a non-JSONRPC payload for an EVM service:
+	//		QoS returns a properly formatted error response.
+	// 	2. Protocol relay failed for any reason:
+	//		QoS returns a generic, properly formatted response: e.g. a JSONRPC error response.
+	//	3. Protocol relay was sent successfully:
+	//		QoS returns the endpoint's response: e.g. the chain ID for a `eth_chainId` request.
+	rc.writeHTTPResponse(rc.qosCtx.GetHTTPResponse(), w)
+}
+
+// writeResponse uses the supplied http.ResponseWriter to write the supplied HTTP response.
+func (rc *requestContext) writeHTTPResponse(response HTTPResponse, w http.ResponseWriter) {
+	for key, value := range response.GetHTTPHeaders() {
+		w.Header().Set(key, value)
+	}
+
+	statusCode := response.GetHTTPStatusCode()
+	// TODO_IMPROVE: introduce handling for cases where the status code is not set.
+	if statusCode == 0 {
+		statusCode = http.StatusOK
+	}
+
+	responsePayload := response.GetPayload()
+	logger := rc.logger.With(
+		"http_response_payload_length", len(responsePayload),
+		"http_response_status", statusCode,
+	)
+
+	w.WriteHeader(statusCode)
+
+	numWrittenBz, writeErr := w.Write(responsePayload)
+	if writeErr != nil {
+		logger.With("http_response_bytes_writte", numWrittenBz).Warn().Err(writeErr).Msg("Error writing the HTTP response.")
+		return
+	}
+
+	logger.Info().Msg("Completed processing the HTTP request and returned an HTTP response.")
+}
+
+// BroadcastAllObservations delivers the collected details regarding all aspects of the service request to all the interested parties.
+// For example:
+//   - QoS-level observations, e.g. endpoint validation results
+//   - Protocol-level observations, e.g. "maxed-out" endpoints.
+//   - Gateway-level observations, e.g. the request ID.
+func (rc *requestContext) BroadcastAllObservations() {
+	var (
+		protocolObservations protocolobservations.Observations
+		qosObservations      qosobservations.Observations
+	)
+
+	// observation-related tasks are called in Goroutines to avoid potentially blocking the HTTP handler.
+	go func() {
+		if rc.protocolCtx != nil {
+			protocolObservations = rc.protocolCtx.GetObservations()
+			err := rc.protocol.ApplyObservations(&protocolObservations)
+			if err != nil {
+				rc.logger.Warn().Err(err).Msg("error publishing protocol observations.")
+			}
+		}
+
+		// The service request context contains all the details the QoS needs to update its internal metrics about endpoint(s), which it should use to build
+		// the observation.QoSObservations struct.
+		// This ensures that separate PATH instances can communicate and share their QoS observations.
+		if rc.qosCtx != nil {
```

Comment:
Can you just do an 

```go
if rc.qosCtx == nil {
	// Log that this is neal and explain why
	return
}
// Avoid indent in the rest of the file
```

Reply:
Agreed in general. But in this specific case, a nil `qosCtx` does not remove the need for the rest of the logic, e.g. calling the `metricsReporter`.